src/sinks/gcp/cloud_storage.rs

```diff
@@ -0,0 +1,543 @@
+            Template::from(kp.as_str())
+        } else {
+            Template::from("date=%F/")
+        };
+
+        let settings = self.settings.clone();
+
+        let svc = ServiceBuilder::new()
+            .map(move |req| RequestWrapper::new(req, settings.clone()))
+            .settings(request, GcsRetryLogic)
+            .service(self);
+
+        let sink = crate::sinks::util::BatchServiceSink::new(svc, cx.acker())
+            .partitioned_batched_with_min(PartitionBuffer::new(Buffer::new(compression)), &batch)
+            .with_flat_map(move |e| iter_ok(encode_event(e, &key_prefix, &encoding)));
+
+        Ok(Box::new(sink))
+    }
+
+    pub fn healthcheck(&self) -> crate::Result<Healthcheck> {
+        let mut builder = Request::builder();
+        builder.method(Method::HEAD);
+        builder.uri(self.base_url.parse::<Uri>()?);
+
+        let mut request = builder.body(Body::empty()).unwrap();
+        if let Some(creds) = &self.creds {
+            creds.apply(&mut request);
+        }
+
+        let healthcheck =
+            self.client
+                .request(request)
+                .map_err(Into::into)
+                .and_then(healthcheck_response(
+                    self.creds.clone(),
+                    GcsError::BucketNotFound {
+                        bucket: self.bucket.clone(),
+                    }
+                    .into(),
+                ));
+
+        Ok(Box::new(healthcheck))
+    }
+}
+
+impl Service<RequestWrapper> for GcsSink {
+    type Response = hyper::Response<Body>;
+    type Error = hyper::Error;
+    type Future = hyper::client::ResponseFuture;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        Ok(().into())
+    }
+
+    fn call(&mut self, request: RequestWrapper) -> Self::Future {
+        let settings = request.settings;
+
+        let uri = format!("{}{}", self.base_url, request.key)
+            .parse::<Uri>()
+            .unwrap();
+        let mut builder = Request::builder();
+        builder.method(Method::PUT);
+        builder.uri(uri);
+        let headers = builder.headers_mut().unwrap();
+        headers.insert("content-type", HeaderValue::from_str("text/plain").unwrap()); // FIXME needs proper content type
+        headers.insert(
+            "content-length",
+            HeaderValue::from_str(&format!("{}", request.body.len())).unwrap(),
+        );
+        settings
+            .content_encoding
+            .map(|ce| headers.insert("content-encoding", ce));
+        headers.insert("x-goog-acl", settings.acl);
+        headers.insert("x-goog-storage-class", settings.storage_class);
+        for (p, v) in settings.tags {
+            headers.insert(p, v);
+        }
+
+        let mut request = builder.body(Body::from(request.body)).unwrap();
+        if let Some(creds) = &self.creds {
+            creds.apply(&mut request);
+        }
+
+        self.client.request(request)
+    }
+}
```

Comment:
This makes me think we could implement this as a `HttpSink`, we would just need to add partitioning support to `BatchedHttpSink` and provide a custom retry logic. I'm curious if this could reduce some of the code here, what do you think?

Reply:
If it works, this would likely be a real win here. I would tend to leave it as part of #1659 though.